### PR TITLE
fix: Batch Ingress updates

### PR DESCRIPTION
**Pull request by astefanutti**
This PR removes the single patch and update requests made during Ingresses reconciliation, so only one update request is performed at the end of each reconciliation.

This reduces the number of conflicting updates, and possibly some missed updates, in addition to simplifying the reconciliation logic.

### CODE REVIEW

File: pkg/reconciler/ingress/ingress.go [hunk 1]
```diff
@@ -107,16 +106,27 @@ func (c *Controller) ensureCertificate(ctx context.Context, ingress *networkingv
...
-	if _, err := c.patchIngress(ctx, ingress, []byte(patch)); err != nil {
-		c.Logger.Error(err, "Failed to patch Ingress", "ingress", ingress)
-		return err
-	}
+	upsertTLS(ingress, controlClusterContext.Host(), controlClusterContext.Name())
 
 	return nil
 }
 
+func upsertTLS(ingress *networkingv1.Ingress, host, secretName string) {
```
Comment (maleck13) [inline pkg/reconciler/ingress/ingress.go:114]: not sure I am following here, where does the the actual update to the ingress happen?

Comment (astefanutti) [reply pkg/reconciler/ingress/ingress.go:114]: It's done here: https://github.com/Kuadrant/kcp-glbc/blob/c2818df6340055c0f958850d7581ecbc09ea9cb1/pkg/reconciler/ingress/controller.go#L138

The general pattern is the reconcile method is passed with a copy of the object, that's mutated, and those changes are applied once the reconcile method return with no error.

**Comment by mikenairn** [comment]
This looks fine to me. Since the pattern seems to be that the outer controller always does the update of the resource being reconciled, we shouldn't be doing any updates to those resources in the reconciler itself. 

Will let @maleck13 approve once he's happy.

File: pkg/reconciler/ingress/ingress.go (hunk 1 above)
Comment (maleck13) [reply pkg/reconciler/ingress/ingress.go:114]: I am just a bit worried with the work I am doing that I will have quite a task rebasing some of this stuff, that said this is a very small change, we might need to work together on the changes I am making to ensure we keep the pattern correct

Comment (astefanutti) [reply pkg/reconciler/ingress/ingress.go:114]: Right, read operations can be done anywhere.

**Pull request merged**